src/index.ts

```diff
@@ -146,48 +146,70 @@ function getNewtype(model: CaseClass): Reader<Ctx, gen.CustomTypeDeclaration> {
+function getCaseClassDeclaration(caseClass: CaseClass): Reader<Ctx, gen.TypeDeclaration | gen.CustomTypeDeclaration> {
+  if (caseClass.isValueClass) {
+    return getNewtype(caseClass);
+  }
   return ask<Ctx>().chain(({ isReadonly }) =>
-    traverseReader(models, model => {
-      if ('isValueClass' in model && model.isValueClass) {
-        return getNewtype(model as CaseClass);
-      }
-      if (model.hasOwnProperty('values')) {
-        const enumClass = model as EnumClass;
-        return reader.of(
-          gen.typeDeclaration(
-            model.name,
-            gen.keyofCombinator(enumClass.values.map((v: any) => v.name), model.name),
-            true,
-            false
-          )
+    traverseReader(caseClass.members, getProperty).map(properties => {
+      const interfaceDecl = gen.typeCombinator(properties, caseClass.name);
+      if (caseClass.typeParams && caseClass.typeParams.length > 0) {
+        const staticParams = caseClass.typeParams.map(p => `${p.name} extends t.Any`).join(', ');
+        const runtimeParams = caseClass.typeParams.map(p => `${p.name}: ${p.name}`).join(', ');
+        const dependencies = interfaceDecl.properties
+          .map(p => gen.printStatic(p.type))
+          .filter(p => !caseClass.typeParams.map(p => p.name).includes(p));
+        return gen.customTypeDeclaration(
+          caseClass.name,
+          `export interface ${caseClass.name}<${caseClass.typeParams.map(p => p.name)}> ${gen.printStatic(
+            interfaceDecl
+          )}`,
+          `export const ${caseClass.name} = <${staticParams}>(${runtimeParams}) => ${gen.printRuntime(interfaceDecl)}`,
+          dependencies
         );
+      } else {
+        return gen.typeDeclaration(caseClass.name, interfaceDecl, true, isReadonly);
       }
-      const caseClass = model as CaseClass;
-      return traverseReader(caseClass.members, getProperty).map(properties => {
-        const interfaceDecl = gen.typeCombinator(properties, model.name);
-        if (caseClass.typeParams && caseClass.typeParams.length > 0) {
-          const staticParams = caseClass.typeParams.map(p => `${p.name} extends t.Any`).join(', ');
-          const runtimeParams = caseClass.typeParams.map(p => `${p.name}: ${p.name}`).join(', ');
-          const dependencies = interfaceDecl.properties
-            .map(p => gen.printStatic(p.type))
-            .filter(p => !caseClass.typeParams.map(p => p.name).includes(p));
-          return gen.customTypeDeclaration(
-            model.name,
-            `export interface ${model.name}<${caseClass.typeParams.map(p => p.name)}> ${gen.printStatic(
-              interfaceDecl
-            )}`,
-            `export const ${model.name} = <${staticParams}>(${runtimeParams}) => ${gen.printRuntime(interfaceDecl)}`,
-            dependencies
-          );
-        } else {
-          return gen.typeDeclaration(model.name, gen.typeCombinator(properties, model.name), true, isReadonly);
```

Comment:
Ha senso che `isReadonly` sia usato solo qui e non quando ci sono parametri di tipo?

(Se sì, eventualmente si può spostare la `ask` nell'`else`)

Reply:
onestamente non credo che `isReadonly` sia utile del tutto. @giogonzo secondo te possiamo levarlo?

Reply:
Confermo, è un mezzo leftover, toglierei isReadonly del tutto

Reply:
"del tutto" intendi da `getCaseClassDeclaration` anche nel caso senza parametri di tipo (mettendolo a `false`) o anche altrove? la `getType` lo usa per scegliere fra `arrayCombinator` e `readonlyArrayCombinator`, quello continua ad avere senso?

Reply:
si intendevo ovunque, nel senso che non abbiamo mai valutato bene ne voluto usare `ReadOnlyArray` ne altri tipi di readonly in TS. Ma se sono modifiche unrelated, piuttosto apriamo un issue e lo facciamo un'altra volta

Reply:
ok, lascio questa così e apro un issue separato per togliere i readonly